crates/sui-graphql-rpc/src/types/dot_move/named_type.rs

```diff
@@ -0,0 +1,203 @@
+    named_move_package::NamedMovePackage,
+};
+
+pub(crate) struct NamedType;
+
+impl NamedType {
+    pub(crate) async fn query(
+        ctx: &Context<'_>,
+        name: &str,
+        checkpoint_viewed_at: u64,
+    ) -> Result<String, Error> {
+        // we do not de-duplicate the names here, as the dataloader will do this for us.
+        let names = Self::parse_names(name)?;
+
+        // Gather all the requests to resolve the names.
+        let names_to_resolve = names
+            .iter()
+            .map(|x| NamedMovePackage::query(ctx, x, checkpoint_viewed_at))
+            .collect::<Vec<_>>();
+
+        // now we resolve all the names in parallel (data-loader will do the proper de-duplication / batching for us)
+        // also the `NamedMovePackage` query will re-validate the names (including max length, which is not checked on the regex).
+        let mut results = future::try_join_all(names_to_resolve).await?;
+
+        // now let's create a hashmap with {name: MovePackage}
+        let mut name_package_id_mapping = HashMap::new();
+
+        // doing it in reverse so we can pop instead of shift
+        for name in names.into_iter().rev() {
+            // safe unwrap: we know that the amount of results has to equal the amount of names.
+            let Some(package) = results.pop().unwrap() else {
+                return Err(Error::DotMove(DotMoveServiceError::NameNotFound(name)));
+            };
+
+            name_package_id_mapping.insert(name, package.native.id());
+        }
+
+        let correct_type_tag = Self::replace_names(name, &name_package_id_mapping);
+
+        // now we query the names with futures to utilize data loader
+        Ok(correct_type_tag)
+    }
+
+    // TODO: Should we introduce some overall string limit length here?
+    // Is this already caught by the global limits?
+    // This parser just extracts all names from a type tag, and returns them
+    // We do not care about de-duplication, as the dataloader will do this for us.
+    // The goal of replacing all of them with `0x0` is to make sure that the type tag is valid
+    // so when replaced with the move name package addresses, it'll also be valid.
+    fn parse_names(name: &str) -> Result<Vec<String>, Error> {
+        let mut names = vec![];
+        let struct_tag = VERSIONED_NAME_UNBOUND_REG.replace_all(name, |m: &regex::Captures| {
+            // safe unwrap: we know that the regex will always have a match on position 0.
+            names.push(m.get(0).unwrap().as_str().to_string());
+            "0x0".to_string()
+        });
+
+        // We attempt to parse the type_tag with these replacements, to make sure there are no other
+        // errors in the type tag (apart from the move names). That protects us from unnecessary
+        // queries to resolve .move names, for a type tag that will be invalid anyway.
+        parse_type_tag(&struct_tag).map_err(|e| Error::Client(e.to_string()))?;
```

Comment:
Looks like we tend to use `TypeTag::from_str` for this in the GraphQL codebase -- let's use that here as well? (It's calling exactly the same code, but keeping it standard will lead to fewer questions about what the difference is).